File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.c
int maxProfit(int* prices, int pricesSize) {
    if (pricesSize <= 1) return 0;

    int min_price = prices[0];
    int max_profit = 0;

    for (int i = 1; i < pricesSize; i++) {
        if (prices[i] < min_price) {
            min_price = prices[i];  // better buying price
        } else {
            int profit = prices[i] - min_price; 
            if (profit > max_profit) {
                max_profit = profit;  // better selling profit
            }
        }
    }

    return max_profit;
}

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int left = 0;  // buy day
        int right = 1; // sell day

        int maxProfit = 0;

        while (right < prices.length) {
            if (prices[right] > prices[left]) {
                int profit = prices[right] - prices[left];
                maxProfit = Math.max(maxProfit, profit);
            } else {
                left = right; // found cheaper buying price
            }

            right++;
        }

        return maxProfit;
    }
}